File: apps/api/plane/db/models/feature_flag.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

from django.db import models
from .base import BaseModel
from .user import User


class FeatureFlag(BaseModel):
    """
    Feature flags at different levels (workspace, project, user)
    """

    class Level(models.TextChoices):
        WORKSPACE = "WORKSPACE", "Workspace"
        PROJECT = "PROJECT", "Project"
        USER = "USER", "User"

    key = models.CharField(max_length=100, unique=True)
    name = models.CharField(max_length=100)
    description = models.TextField(blank=True)
    level = models.CharField(
        max_length=20,
        choices=Level.choices,
        default=Level.WORKSPACE,
    )
    is_default = models.BooleanField(default=False)
    default_enabled = models.BooleanField(default=False)

    class Meta:
        app_label = "db"
        verbose_name = "Feature Flag"
        verbose_name_plural = "Feature Flags"

    def __str__(self):
        return f"{self.name} ({self.key})"


class FeatureFlagValue(BaseModel):
    """
    Stores feature flag values for workspaces, projects, and users
    """

    class Level(models.TextChoices):
        WORKSPACE = "WORKSPACE", "Workspace"
        PROJECT = "PROJECT", "Project"
        USER = "USER", "User"

    feature_flag = models.ForeignKey(
        FeatureFlag,
        on_delete=models.CASCADE,
        related_name="values",
    )
    level = models.CharField(
        max_length=20,
        choices=Level.choices,
    )
    entity_id = models.CharField(max_length=100)
    enabled = models.BooleanField(default=False)

    workspace = models.ForeignKey(
        "db.Workspace",
        on_delete=models.CASCADE,
        null=True,
        blank=True,
        related_name="feature_flags",
    )
    project = models.ForeignKey(
        "db.Project",
        on_delete=models.CASCADE,
        null=True,
        blank=True,
        related_name="feature_flags",
    )
    user = models.ForeignKey(
        User,
        on_delete=models.CASCADE,
        null=True,
        blank=True,
        related_name="feature_flags",
    )

    class Meta:
        app_label = "db"
        verbose_name = "Feature Flag Value"
        verbose_name_plural = "Feature Flag Values"
        unique_together = ["feature_flag", "level", "entity_id"]
        indexes = [
            models.Index(fields=["level", "entity_id"]),
            models.Index(fields=["workspace", "level"]),
            models.Index(fields=["project", "level"]),
            models.Index(fields=["user", "level"]),
        ]

    def __str__(self):
        return f"{self.feature_flag.name} - {self.level}:{self.entity_id} = {self.enabled}"

File: packages/types/src/feature-flag.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

export enum EFeatureFlag {
  AI_ASSISTANT = "ai_assistant",
  EXPORT = "export",
  IMPORT = "import",
  ANALYTICS = "analytics",
  GITHUB_IMPORT = "github_import",
  JIRA_IMPORT = "jira_import",
  CUSTOM_ROLES = "custom_roles",
  WEBHOOKS = "webhooks",
  API_TOKENS = "api_tokens",
  SLA = "sla",
  ISSUE_AUTOMATIONS = "issue_automations",
  WORKFLOW_AUTOMATIONS = "workflow_automations",
}

export type TFeatureFlag = EFeatureFlag;

export enum EFeatureFlagLevel {
  WORKSPACE = "WORKSPACE",
  PROJECT = "PROJECT",
  USER = "USER",
}

export type TFeatureFlagLevel = EFeatureFlagLevel;

export interface IFeatureFlag {
  id: string;
  key: EFeatureFlag;
  name: string;
  description: string;
  level: TFeatureFlagLevel;
}

export interface IFeatureFlagValue {
  workspace: Record<string, boolean>;
  project: Record<string, Record<string, boolean>>;
  user: Record<string, Record<string, boolean>>;
}

export interface IFeatureFlagsResponse {
  flags: IFeatureFlag[];
  values: IFeatureFlagValue;
}

export interface IWorkspaceFeatureFlags {
  [workspaceSlug: string]: Record<EFeatureFlag, boolean>;
}

export interface IProjectFeatureFlags {
  [workspaceSlug: string]: {
    [projectId: string]: Record<EFeatureFlag, boolean>;
  };
}

export interface IUserFeatureFlags {
  [workspaceSlug: string]: {
    [userId: string]: Record<EFeatureFlag, boolean>;
  };
}

export interface IFeatureFlagInput {
  feature_flag: EFeatureFlag;
  level: TFeatureFlagLevel;
  entity_id: string;
  enabled: boolean;
}
